contracts/access/SignatureBouncer.sol

```diff
@@ -20,12 +20,22 @@ import "../ECRecovery.sol";
  * @dev Then restrict access to your crowdsale/whitelist/airdrop using the
  * @dev `onlyValidSignature` modifier (or implement your own using isValidSignature).
  * @dev
+ * @dev In addition to `onlyValidSignature`, `onlyValidSignatureAndMethod` and
+ * @dev `onlyValidSignatureAndData` can be used to restrict access to only a given method
+ * @dev or a given method with given parameters respectively.
+ * @dev
  * @dev See the tests Bouncer.test.js for specific usage examples.
+ * @notice A method that uses the `onlyValidSignatureAndData` modifier must make the _sig
+ * @notice parameter the "last" parameter. You cannot sign a message that has its own
+ * @notice signature in it so the last 160 bytes is of msg.data (which represents _sig)
+ * @notice is ignored when validating.
  */
 contract SignatureBouncer is Ownable, RBAC {
   using ECRecovery for bytes32;
 
   string public constant ROLE_BOUNCER = "bouncer";
+  uint constant METHOD_ID_SIZE = 4;
+  uint constant SIG_SIZE = 160;
```

Comment:
@shrugs I forgot that I wanted to comment on this variable. The signature value is always 160 bytes within `msg.data` even though `_sig.length` is always much smaller (64 bytes I think?). I guess it pads this data to the max size? Regardless, it's always correct but I wanted to call attention to the fact that I wasn't totally sure why.

Reply:
I spent a bit of time playing around with the abi documentation and realized that `bytes` are encoded as follows

```
[distance to the beginning of the argument data] -> 32 bytes
[the boolean true] -> 32 bytes
[length of the data] -> 32 bytes
[the data, right-padded to a multiple of 32] -> 64 bytes
```

and there we have our `160`.

---

I have a question, though, after looking at the abi documentation: can we add a test case to see if it's possible to have a dynamic argument _before_ the final `bytes` argument? i.e. can we sign a bouncer signature for 

```solidity
function test(bytes _someBytes, bytes _sig)
   onlyValidSignatureAndData(_sig)
{
}
```

my hunch is that this won't work immediately because with dynamically sized arguments, their location in the transaction data is included in the tx data _before_ the argument itself.

i.e. I think a transaction for the above function looks like

```
0x12345678 // 8 bytes
  [ location of data for _someBytes ] // (8 bytes + 32 + 32)
  [ location of data for _sig ] // (8 + 32 + 32 + 32 + len(_someBytes))
  [ enc(len(_someBytes) ]
  [ _someBytes right-padded to modulo 32 ]
  [ enc(len(_sig)) ]
  [ _sig right-padded to module 32 ] // (64 bytes long in our case)
```

which means we're lucky that the examples only use fixed-sized arguments so we don't have to calculate stuff like "at what offset will the encoded signature be in the transaction data for this function call".

Reply:
Thanks for shedding some light on this. I'll add that test and see what we get